### PR TITLE
[Clang] use begin member expr location for call expr with deducing this

### LINKED ISSUE #116928: CallExpr has incorrect begin location with "deducing this"
In the following code:

```c++
struct S
{
    int bar(this S&);
};

int main()
{
    S waldo;
    int x = waldo.bar();
}
```

The AST dump of `main()`, taken with a recent trunk, is:

```
`-FunctionDecl 0x55ac3a911f78 <line:6:1, line:10:1> line:6:5 main 'int ()'
  `-CompoundStmt 0x55ac3a912988 <line:7:1, line:10:1>
    |-DeclStmt 0x55ac3a9127c8 <line:8:5, col:12>
    | `-VarDecl 0x55ac3a9120a8 <col:5, col:7> col:7 used waldo 'S' callinit
    |   `-CXXConstructExpr 0x55ac3a9127a0 <col:7> 'S' 'void () noexcept'
    `-DeclStmt 0x55ac3a912970 <line:9:5, col:24>
      `-VarDecl 0x55ac3a9127f8 <col:5, col:23> col:9 x 'int' cinit
        `-CallExpr 0x55ac3a912948 <col:19, col:23> 'int'
          |-ImplicitCastExpr 0x55ac3a912930 <col:19> 'int (*)(S &)' <FunctionToPointerDecay>
          | `-DeclRefExpr 0x55ac3a9128b0 <col:19> 'int (S &)' lvalue CXXMethod 0x55ac3a911e58 'bar' 'int (S &)'
          `-DeclRefExpr 0x55ac3a912860 <col:13> 'S' lvalue Var 0x55ac3a9120a8 'waldo' 'S'
```

The `CallExpr` for `waldo.bar()` is reporting `col:19` as its begin location, even though the descendant `DeclRefExpr` for `waldo` is at `col:13`.

This is misleading, and causes problems for tooling like clangd whose hit-testing algorithm (used for go-to-definition) skips the `CallExpr` node when hit-testing the `waldo` token because it falls outside of the `CallExpr`'s reported source range (https://github.com/llvm/llvm-project/issues/116335).

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #116928

### CODE REVIEW

**Review by zyn0217 — APPROVED** [review]
LGTM modulo a release note.

**Comment by a-tarasyuk** [comment]
@zyn0217 thanks for the review. I've added a release note.

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -717,6 +717,8 @@ Bug Fixes to C++ Support
 - Clang now uses valid deduced type locations when diagnosing functions with trailing return type
   missing placeholder return type. (#GH78694)
 - Fixed a bug where bounds of partially expanded pack indexing expressions were checked too early. (#GH116105)
+- Clang now uses the location of the begin of the member expression for ``CallExpr``
```
Comment (zyn0217) [inline clang/docs/ReleaseNotes.rst:720]: It would fit better under `bug fixes to ast handling`, right?

Comment (a-tarasyuk) [reply clang/docs/ReleaseNotes.rst]: @zyn0217 I've moved to `bug fixes to ast handling`

**Review by cor3ntin — APPROVED** [review]

**Review by zyn0217 — APPROVED** [review]

**Pull request merged**